app/jobs/data_import_job.rb

```diff
@@ -6,11 +6,24 @@ class DataImportJob < ApplicationJob
 
   def perform(data_import)
     contacts = []
-    data_import.update!(status: :processing)
-    csv = CSV.parse(data_import.import_file.download, headers: true)
-    csv.each { |row| contacts << build_contact(row.to_h.with_indifferent_access, data_import.account) }
-    result = Contact.import contacts, on_duplicate_key_update: :all, batch_size: 1000
-    data_import.update!(status: :completed, processed_records: csv.length - result.failed_instances.length, total_records: csv.length)
+    rejected_contacts = []
+    @data_import = data_import
+    @data_import.update!(status: :processing)
+    csv = CSV.parse(@data_import.import_file.download, headers: true)
+
+    csv.each do |row|
+      current_contact = build_contact(row.to_h.with_indifferent_access, @data_import.account)
+      current_contact.valid? ? contacts << current_contact : rejected_contacts << current_contact
+    end
+
+    # <struct ActiveRecord::Import::Result failed_instances=[], num_inserts=1, ids=[444, 445], results=[]>
+    Contact.import(contacts, synchronise: contacts, on_duplicate_key_ignore: true, track_validation_failures: true, validate: true,
```

Comment:
Note: I will take a pass at `failed_instances` and see if we can get it populated.